apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs

```diff
@@ -0,0 +1,172 @@
+            ConnectionString = connectionString;
+        }
+
+        /// <summary>
+        /// </summary>
+        public SpannerConnectionStringBuilder()
+        {
+        }
+
+        /// <summary>
+        /// </summary>
+        public string ClientId
+        {
+            get { return (string) this[nameof(ClientId)]; }
+            private set { this[nameof(ClientId)] = value; }
+        }
+
+        /// <summary>
+        /// </summary>
+        public string ClientSecret
+        {
+            get { return (string) this[nameof(ClientSecret)]; }
+            private set { this[nameof(ClientSecret)] = value; }
+        }
+
+        /// <summary>
+        /// </summary>
+        public ITokenAccess Credential { get; private set; }
+
+        /// <summary>
+        /// </summary>
+        public string DataSource
+        {
+            get { return (string) this["Data Source"]; }
+            private set { this["Data Source"] = value; }
+        }
+
+        /// <summary>
+        /// </summary>
+        public ServiceEndpoint EndPoint => new ServiceEndpoint(Host, Port);
+
+        /// <summary>
+        /// </summary>
+        public string Host
+        {
+            get { return (string) this["Host"] ?? SpannerClient.DefaultEndpoint.Host; }
+            private set { this["Host"] = value; }
+        }
+
+        /// <summary>
+        /// </summary>
+        public int Port
+        {
+            get
+            {
+                var result = SpannerClient.DefaultEndpoint.Port;
+                var value = (string) this["Port"];
+                if (value != null)
+                    if (!int.TryParse(value, out result))
+                        result = SpannerClient.DefaultEndpoint.Port;
+                return result;
+            }
+            private set { this["Port"] = value.ToString(); }
+        }
+
+        /// <summary>
+        /// </summary>
+        public string Project => ParsedDataSourcePart(0);
+
+        /// <summary>
+        /// </summary>
+        public string SpannerDatabase => ParsedDataSourcePart(2);
+
+        /// <summary>
+        /// </summary>
+        public string SpannerInstance => ParsedDataSourcePart(1);
+
+        /// <summary>
+        /// </summary>
+        public string UserName
+        {
+            get { return (string) this[nameof(UserName)]; }
+            private set { this[nameof(UserName)] = value; }
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <returns></returns>
+        public async Task<UserCredential> QueryForCredentials()
+        {
+            if (Credential == null)
+            {
+                var clientId = ClientId;
+                var clientSecret = ClientSecret;
+                if (!string.IsNullOrEmpty(clientId)
+                    && !string.IsNullOrEmpty(clientSecret))
+                    return
+                        await
+                            GoogleWebAuthorizationBroker.AuthorizeAsync(
```

Comment:
I think if we want to do client-side authentication, we shouldn't assume it would be using `GoogleWebAuthorizationBroker`. But let's talk to the spanner folks about what auth options they actually require.

Personally I'd have expected the `ITokenAccess` specified to be the one to use for the channel credentials, and not to use anything for the `CallCredentials`. If the user wants to do client access, they can do the auth dance however they want to, and then provide that as the `ITokenAccess`. If it proves useful to have separate call credentials, we could potentially support that as a second auth token - but it sounds overly complicated for now.